### PR TITLE
fix: update CR's status with retry.RetryOnConflict

### PR DESCRIPTION
**Pull request by dylandee**
It's would be better to update CR's status with retry.RetryOnConflict to avoid the conflicts caused by other code making unrelated updates to the resource at the same.

### CODE REVIEW

**Comment by wonderflow** [comment]
Can you describe more detail about:

> the conflicts caused by other code making unrelated updates to the resource at the same.

**Comment by dylandee** [comment]
> Can you describe more detail about:
> 
> > the conflicts caused by other code making unrelated updates to the resource at the same.
> 
> Update with retry blindly will gloss over problems some of which we should notice and fix.

As what is used in client-go examples(https://github.com/kubernetes/client-go/blob/93ce9718ffcde5a4465be99b1e0238a3a042ad22/examples/create-update-delete-deployment/main.go#L126)

this function is  used to retry status update only if CR's version is not the latest and other errors wont be missed if raised

Waht's more, this kind of function is widely used like in OpenKruise
https://github.com/openkruise/kruise/blob/faa8045ae74b1f2d57a173ea22fa1439580580e7/pkg/controller/advancedcronjob/advancedcronjob_controller.go#L165

**Comment by wonderflow** [comment]
Yes, the example in openkruise makes sense, can you help make the function to be a method of those reconciler and name it like update<XXX>Status?

I don't want the function to be abused

**Comment by dylandee** [comment]
> Yes, the example in openkruise makes sense, can you help make the function to be a method of those reconciler and name it like updateStatus?
> 
> I don't want the function to be abused

I have tried to update status with one function at first time, but i find the function would be complex since i have to use 'switch...case...' to distinguish different CR's struct. 
Is it better to define one status update function in every  reconciler file?

**Comment by wonderflow** [comment]
> Is it better to define one status update function in every reconciler file?

Yes, of course. You could write one for each reconciler

**Comment by dylandee** [comment]
> Yes, the example in openkruise makes sense, can you help make the function to be a method of those reconciler and name it like updateStatus?
> 
> I don't want the function to be abused

Done.

**Review by wonderflow — APPROVED** [review]
Great Job!  LGTM

ping @ryanzhang-oss @hongchaodeng  @captainroy-hy  would you like to have a second review?
The status update function widely changed our old way, I'll leave it until next Monday before merge.

**Review by captainroy-hy — APPROVED** [review]
LGTM.

File: pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/component.go
Comment (ryanzhang-oss) [inline pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/component.go:284]: what exactly does this function do?

File: pkg/controller/core.oam.dev/v1alpha2/core/scopes/healthscope/healthscope_controller.go
Comment (ryanzhang-oss) [inline pkg/controller/core.oam.dev/v1alpha2/core/scopes/healthscope/healthscope_controller.go:264]: is there a way to not repeat this over and over?

Comment (wonderflow) [reply pkg/controller/core.oam.dev/v1alpha2/core/scopes/healthscope/healthscope_controller.go:264]: That's because Go don't have the generic.

Comment (ryanzhang-oss) [reply pkg/controller/core.oam.dev/v1alpha2/core/scopes/healthscope/healthscope_controller.go:264]: we could use a common function

**Pull request merged**
